flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/action/DeleteAction.java

```diff
@@ -0,0 +1,183 @@
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.store.connector.action.Action.getTablePath;
+import static org.apache.flink.table.store.file.catalog.Catalog.DEFAULT_DATABASE;
+
+/** Delete from table action for Flink. */
+public class DeleteAction extends AbstractActionBase {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DeleteAction.class);
+
+    private final FlinkCatalog flinkCatalog;
+
+    @Nullable private String filter;
+
+    public DeleteAction(Path tablePath) {
+        super(tablePath);
+        flinkCatalog = new FlinkCatalog(catalog, "table-store", DEFAULT_DATABASE);
+    }
+
+    public DeleteAction withFilter(String filter) {
+        this.filter = filter;
+        return this;
+    }
+
+    public static Optional<Action> create(String[] args) {
+        LOG.info("Delete job args: {}", String.join(" ", args));
+
+        MultipleParameterTool params = MultipleParameterTool.fromArgs(args);
+
+        if (params.has("help")) {
+            printHelp();
+            return Optional.empty();
+        }
+
+        Path tablePath = getTablePath(params);
+
+        if (tablePath == null) {
+            return Optional.empty();
+        }
+
+        DeleteAction action = new DeleteAction(tablePath);
+
+        if (params.has("where")) {
+            String filter = params.get("where");
+            if (filter == null) {
+                return Optional.empty();
+            }
+
+            action.withFilter(filter);
+        }
+
+        return Optional.of(action);
+    }
+
+    private static void printHelp() {
+        System.out.println("Action \"delete\" deletes data from a table.");
+        System.out.println();
+
+        System.out.println("Syntax:");
+        System.out.println(
+                "  delete --warehouse <warehouse-path> --database <database-name> "
+                        + "--table <table-name> [--where <filter_spec>]");
+        System.out.println("  delete --path <table-path> [--where <filter_spec>]");
+        System.out.println();
+
+        System.out.println(
+                "The '--where <filter_spec>' part is equal to the 'WHERE' clause in SQL DELETE statement");
+        System.out.println();
+
+        System.out.println("Examples:");
+        System.out.println(
+                "  delete --warehouse hdfs:///path/to/warehouse --database test_db --table test_table");
+        System.out.println("  It's equal to 'DELETE FROM test_table' which deletes all records");
+        System.out.println();
+        System.out.println(
+                "  delete --path hdfs:///path/to/warehouse/test_db.db/test_table --where id > (SELECT count(*) FROM employee)");
+        System.out.println(
+                "  It's equal to 'DELETE FROM test_table WHERE id > (SELECT count(*) FROM employee)");
+    }
+
+    @Override
+    public void run() throws Exception {
+        if (filter == null) {
+            LOG.debug("Run delete action with no filter.");
```

Comment:
I think we can throw exception here to alert users to use overwrite.
`table.deleteWhere` has performance problem if the table is very large.